### PR TITLE
refactor: serialize extra json in state

### PR DESCRIPTION
**Pull request by eschutho**
### SUMMARY
refactoring the extra json state in the database connection modal so that it is always serialized, as the api will expect it. Each component will deserialize and push it back to the reducer, which will then serialize it again. This should reduce some duplication of types and state. 

This is also in preparation for adding a databricks form which will need to have extra json property as a field.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
No visual changes

### TESTING INSTRUCTIONS
Adding and removing extra json in the advanced section should work as expected

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by geido** [comment]
@eschutho when opening the "Other" tab under "Advanced" the page crashes. I checked it on master and it is working fine, the bug appears only on the ephemeral env. I think it is worth having a look

**Comment by eschutho** [comment]
> @eschutho when opening the "Other" tab under "Advanced" the page crashes. I checked it on master and it is working fine, the bug appears only on the ephemeral env. I think it is worth having a look

Thanks for testing! I'll take a look.

**Review by betodealmeida — APPROVED** [review]
This looks amazing and a lot of work — thanks for refactoring this!

**Comment by andrey-zayats** [comment]
The "Allow this database to be explored" setting should be checked by default in the Database Connection UI. I can't repro the issue in latest master.

https://user-images.githubusercontent.com/90777564/192780216-1ef52100-541c-4c47-9d6e-2db69c2afac6.mp4

**Comment by andrey-zayats** [comment]
User can't create a dataset using newly connected gsheet.

https://user-images.githubusercontent.com/90777564/192787621-d8ee0ab9-31dc-46c4-8ae0-1bc8145af6e5.mp4

File: superset-frontend/src/views/CRUD/data/database/DatabaseModal/index.tsx [hunk 1]
```diff
@@ -271,26 +276,44 @@ function dbReducer(
...
-        const catalogToUpdate = trimmedState?.catalog[idx] || {};
-        catalogToUpdate[action.payload.name] = action.payload.value;
-
-        const paramatersCatalog = {};
-        // eslint-disable-next-line array-callback-return
-        trimmedState.catalog?.map((item: CatalogObject) => {
-          paramatersCatalog[item.name] = item.value;
-        });
-
+      if (action.payload.type?.startsWith('catalog')) {
```
Comment (AAfghahi) [inline superset-frontend/src/views/CRUD/data/database/DatabaseModal/index.tsx:279]: why have this be a nested loop?

Comment (eschutho) [reply superset-frontend/src/views/CRUD/data/database/DatabaseModal/index.tsx]: Thanks for calling that out. I had originally written logic for the case where `action.payload.type?.startsWith('catalog') && trimmedState.catalog === undefined`, thus the nested logic, but now I see that a form with a catalog will always have a defined catalog state, so I removed the logic and added some comments.

**Pull request merged**